druid/src/text/layout.rs

```diff
@@ -0,0 +1,257 @@
+    }
+
+    /// Given the utf-8 position of a character boundary in the underlying text,
+    /// return a `Line` suitable for drawing a vertical cursor at that boundary.
+    pub fn cursor_line_for_text_position(&self, text_pos: usize) -> Line {
+        self.layout
+            .as_ref()
+            .map(|layout| {
+                let pos = layout.hit_test_text_position(text_pos);
+                let line_metrics = layout.line_metric(pos.line).unwrap();
+                let p1 = (pos.point.x, line_metrics.y_offset);
+                let p2 = (pos.point.x, (line_metrics.y_offset + line_metrics.height));
+                Line::new(p1, p2)
+            })
+            .unwrap_or_else(|| Line::new(Point::ZERO, Point::ZERO))
+    }
+
+    /// Called during the containing widgets `update` method; this text object
+    /// will check to see if any used environment items have changed,
+    /// and invalidate itself as needed.
+    ///
+    /// Returns `true` if an item has changed, indicating that the text object
+    /// needs layout.
+    ///
+    /// # Note
+    ///
+    /// After calling this method, the layout may be invalid until the next call
+    /// to [`rebuild_layout_if_needed`], [`layout`], or [`paint`].
+    ///
+    /// [`layout`]: #method.layout
+    /// [`paint`]: #method.paint
+    /// [`rebuild_layout_if_needed`]: #method.rebuild_layout_if_needed
+    pub fn rebuild_if_needed(&mut self, factory: &mut PietText, text: &str, env: &Env) {
+        let new_font = self.font.resolve(env);
+        let new_color = self.text_color.resolve(env);
+        let new_size = self.text_size_override.as_ref().map(|key| key.resolve(env));
+        let text_changed = self.layout.as_ref().map(|layout| layout.text()) != Some(text);
+
+        let needs_rebuild = !new_font.same(&self.cached_font)
+            || !new_color.same(&self.cached_text_color)
+            || new_size != self.cached_text_size
+            || text_changed;
+
+        self.cached_font = new_font;
+        self.cached_text_color = new_color;
+        self.cached_text_size = new_size;
+
+        if needs_rebuild {
+            let descriptor = if let Some(size) = &self.cached_text_size {
+                self.cached_font.clone().with_size(*size)
+            } else {
+                self.cached_font.clone()
+            };
+            let text_color = self.cached_text_color.clone();
+            self.layout = Some(
+                factory
+                    .new_text_layout(text)
+                    .font(descriptor.family.clone(), descriptor.size)
+                    .default_attribute(descriptor.weight)
+                    .default_attribute(descriptor.style)
+                    .default_attribute(TextAttribute::ForegroundColor(text_color))
+                    .build()
+                    .unwrap(),
+            )
+        }
+    }
+
+    ///  Draw the layout at the provided `Point`.
+    ///
+    ///  You must call [`rebuild_if_needed`] at some point before you first
```

Comment:
This is probably a good place to mention that the point references the top left of the text layout area.